### DISCUSSION TITLE
Tracking `GetAllocatedBytesForCurrentThread` in `async` method

### DISCUSSION THREAD

**Discussion by lsoft** [dotnet/runtime#71530] [Q&A]
I need to collect allocations for a method (or for hierarchy of methods) with `await`s. Something like the following:

```C#
//code not working actually!
var alloc1 = GC.GetAllocatedBytesForCurrentThread();

var mem = new byte[1024];
await Task.Delay(100);

var alloc2 = GC.GetAllocatedBytesForCurrentThread(); //may be other, unrelated, value, due to different thread
var diff = alloc2 - alloc1;
Console.WriteLine("Allocated bytes: " + diff);
```

I can't use global allocation counters, due to a lot of threads runs in the same time, and I need to know allocation for all of them individually.

I tried custom `AsyncTaskMethodBuilder`: 1) aggregate allocations between thread switching into some kind of `AsyncLocal` container 2) and restart to collect after thread switching, but looks like that's infeasible, at least with my way.

Any ideas? Approaches?

Thank you!

**Comment by EgorBo** [comment]
Is this a console app? If you have a sync context, await should restore you back to the original thread. GC.GetAllocatedBytesForCurrentThread for threadpool'ed thread probably doesn't make much sense

**Reply by lsoft** [reply]
@EgorBo in my case it is windows service, but would be cool to have a solution that works in any kind of application. also scheduling to the original thread does not work - while my code awaiting, this thread may execute other code which allocate (if I understand things correctly), so the allocation data will be incorrect.

**Comment by lsoft** [comment]
to clarify: I need any way to track allocations. if it feasible without GetAllocatedBytesForCurrentThread, I'm fine!
